File: src/nodejs/hprose/server/HproseService.js
/**********************************************************\
|                                                          |
|                          hprose                          |
|                                                          |
| Official WebSite: http://www.hprose.com/                 |
|                   http://www.hprose.net/                 |
|                   http://www.hprose.org/                 |
|                                                          |
\**********************************************************/

/**********************************************************\
 *                                                        *
 * HproseService.js                                       *
 *                                                        *
 * HproseService for Node.js.                             *
 *                                                        *
 * LastModified: Dec 28, 2013                             *
 * Author: Ma Bingyao <andot@hprfc.com>                   *
 *                                                        *
\**********************************************************/

var util = require('util');
var EventEmitter = require('events').EventEmitter;
var HproseResultMode = require('../common/HproseResultMode.js');
var HproseException = require('../common/HproseException.js');
var HproseFilter = require('../common/HproseFilter.js');
var HproseTags = require('../io/HproseTags.js');
var HproseBufferInputStream = require('../io/HproseBufferInputStream.js');
var HproseBufferOutputStream = require('../io/HproseBufferOutputStream.js');
var HproseSimpleReader = require('../io/HproseSimpleReader.js');
var HproseSimpleWriter = require('../io/HproseSimpleWriter.js');
var HproseReader = require('../io/HproseReader.js');
var HproseWriter = require('../io/HproseWriter.js');

function callService(method, obj, context, args) {
    var result;
    if (typeof(method) == "function") {
        result = method.apply(context, args);
    }
    else if (obj && typeof(obj[method]) == "function") {
        result = obj[method].apply(context, args);
    }
    return result;
}

function arrayValues(obj) {
    var result = [];
    for (var key in obj) result.push(obj[key]);
    return result;
}

function getFuncName(func, obj) {
    var f = func.toString();
    var funcname = f.substr(0, f.indexOf('(')).replace(/(^\s*function\s*)|(\s*$)/ig, '');
    if ((funcname == "") && obj) {
        for (var name in obj) {
            if (obj[name] === func) return name;
        }
    }
    return funcname;
}

function responseEnd(ostream, response, filter) {
    response.emit("end", filter.outputFilter(ostream.toBuffer()));
}

function getCallback(functionName, functionArgs, byref, resultMode, async, writer, request, response, filter) {
    return function(result) {
        this.emit('afterInvoke', functionName, functionArgs, byref, result, request);
        var ostream = writer.stream;
        if (resultMode == HproseResultMode.RawWithEndTag) {
            ostream.write(result);
            responseEnd(ostream, response, filter);
            return true;
        }
        else if (resultMode == HproseResultMode.Raw) {
            ostream.write(result);
        }
        else {
            ostream.write(HproseTags.TagResult);
            if (resultMode == HproseResultMode.Serialized) {
                ostream.write(result);
            }
            else {
                writer.reset();
                writer.serialize(result);
            }
            if (byref) {
                ostream.write(HproseTags.TagArgument);
                writer.reset();
                writer.writeList(functionArgs);
            }
        }
        if (async) {
            ostream.write(HproseTags.TagEnd);
            responseEnd(ostream, response, filter);
        }
        return false;
    };
}

function HproseService() {
    var m_functions = {};
    var m_funcNames = {};
    var m_debug = false;
    var m_simple = false;
    var m_filter = new HproseFilter();

    EventEmitter.call(this);

    // private methods
    doInvoke = function(istream, request, response) {
        var async = false;
        var ostream = new HproseBufferOutputStream();
        var simpleReader = new HproseSimpleReader(istream);
        do {
            var functionName = simpleReader.readString();
            var aliasName = functionName.toLowerCase();
            var func = m_functions[aliasName] || m_functions['*'];
            var simple, writer;
            if (func) {
                simple = (func.simple === undefined ? m_simple : func.simple);
                writer = (simple ? new HproseSimpleWriter(ostream) : new HproseWriter(ostream));
            }
            else {
                throw new HproseException("Can't find this function " + functionName + "().");
            }
            var functionArgs = [];
            var byref = false;
            var tag = simpleReader.checkTags([HproseTags.TagList,
                                              HproseTags.TagEnd,
                                              HproseTags.TagCall]);
            if (tag == HproseTags.TagList) {
                var reader = new HproseReader(istream);
                functionArgs = reader.readListWithoutTag();
                tag = reader.checkTags([HproseTags.TagTrue,
                                        HproseTags.TagEnd,
                                        HproseTags.TagCall]);
                if (tag == HproseTags.TagTrue) {
                    byref = true;
                    tag = reader.checkTags([HproseTags.TagEnd,
                                            HproseTags.TagCall]);
                }
            }
            this.emit('beforeInvoke', functionName, functionArgs, byref, request);
            var callback = getCallback(functionName, functionArgs, byref, func.resultMode, func.async, writer, request, response, m_filter).bind(this);
            if (func === m_functions['*']) {
                functionArgs = [functionName, functionArgs];
            }
            if (func.async) {
                async = true;
                callService(func.method, func.obj, func.context, functionArgs.concat([callback]));
            }
            else {
                if (callback(callService(func.method, func.obj, func.context, functionArgs))) return;
            }
        } while (tag == HproseTags.TagCall);
        if (!async) {
            ostream.write(HproseTags.TagEnd);
            responseEnd(ostream, response, m_filter);
        }
    }
    
    // protected methods

    this._doFunctionList = function(response) {
        var ostream = new HproseBufferOutputStream();
        var writer = new HproseSimpleWriter(ostream);
        var functions = arrayValues(m_funcNames);
        ostream.write(HproseTags.TagFunctions);
        writer.writeList(functions);
        ostream.write(HproseTags.TagEnd);
        responseEnd(ostream, response, m_filter);
    }

    this._handle = function(data, request, response) {
        var data = m_filter.inputFilter(data);
        var istream = new HproseBufferInputStream(data);
        try {
            switch (istream.getc()) {
                case HproseTags.TagCall: return doInvoke.call(this, istream, request, response);
                case HproseTags.TagEnd: return this._doFunctionList(response);
                default: throw new HproseException("Wrong Request: \r\n" + data.toString());
            }
        }
        catch (e) {
            this.emit('sendError', e);
            var ostream = new HproseBufferOutputStream();
            var writer = new HproseSimpleWriter(ostream);
            ostream.write(HproseTags.TagError);
            writer.writeString(e.message);
            ostream.write(HproseTags.TagEnd);
            responseEnd(ostream, response, m_filter);
        }
    }

    // public methods
    this.isDebugEnabled = function() {
        return m_debug;
    }
    
    this.setDebugEnabled = function(enable) {
        if (enable === undefined) enable = true;
        m_debug = enable;
    }

    this.getSimpleMode = function() {
        return m_simple;
    }
    
    this.setSimpleMode = function(value) {
        if (value === undefined) value = true;
        m_simple = value;
    }

    this.addMissingFunction = function(func, resultMode, async, simple) {
        this.addFunction(func, "*", resultMode, async, simple);
    }
    
    this.addAsyncMissingFunction = function(func, resultMode, simple) {
        this.addMissingFunction(func, resultMode, true, simple);
    }

    this.addMissingMethod = function(method, obj, context, resultMode, async, simple) {
        this.addMethod(method, obj, "*", context, resultMode, async, simple);
    }

    this.addAsyncMissingMethod = function(method, obj, context, resultMode, simple) {
        this.addMissingMethod(method, obj, context, resultMode, true, simple);
    }

    this.addFunction = function(func, alias, resultMode, async, simple) {
        if (resultMode === undefined) {
            resultMode = HproseResultMode.Normal;
        }
        if (alias === undefined || alias == null) {
            switch(typeof(func)) {
                case "string":
                    alias = func;
                    break;
                case "function":
                    alias = getFuncName(func);
                    if (alias != "") break;
                default:
                    throw new HproseException('Need an alias');                            
            }
        }
        if (typeof(alias) == "string") {
            var aliasName = alias.toLowerCase();
            m_functions[aliasName] = {method: func, obj: null, context: null, resultMode: resultMode, async: async, simple: simple};
            m_funcNames[aliasName] = alias;
        }
        else {
            throw new HproseException('Argument alias is not a string');
        }
    }

    this.addAsyncFunction = function(func, alias, resultMode, simple) {
        this.addFunction(func, alias, resultMode, true, simple);
    }

    this.addFunctions = function(functions, aliases, resultMode, async, simple) {
        var count = functions.length;
        var i;
        if (aliases === undefined || aliases == null) {
            for (i = 0; i < count; i++) this.addFunction(functions[i], null, resultMode, async, simple);
        }
        else {
            if (count != aliases.length) {
                throw new HproseException('The count of functions is not matched with aliases');
            }
            for (i = 0; i < count; i++) this.addFunction(functions[i], aliases[i], resultMode, async, simple);
        }
    }

    this.addAsyncFunctions = function(functions, aliases, resultMode, simple) {
        this.addFunctions(functions, aliases, resultMode, true, simple);
    }

    this.addMethod = function(method, obj, alias, context, resultMode, async, simple) {
        if (obj === undefined || obj == null) {
            this.addFunction(method, alias, resultMode, async, simple);
            return;
        }
        if (context === undefined) {
            context = obj;
        }
        if (resultMode === undefined) {
            resultMode = HproseResultMode.Normal;
        }
        if (alias === undefined || alias == null) {
            switch(typeof(method)) {
                case "string":
                    alias = method;
                    break;
                case "function": 
                    alias = getFuncName(method, obj);
                    if (alias != "") break;
                default:
                    throw new HproseException('Need an alias');                            
            }
        }
        if (typeof(alias) == "string") {
            var aliasName = alias.toLowerCase();
            m_functions[aliasName] = {method: method, obj: obj, context: context, resultMode: resultMode, async: async, simple: simple};
            m_funcNames[aliasName] = alias;
        }
        else {
            throw new HproseException('Argument alias is not a string');
        }
    }

    this.addAsyncMethod = function(method, obj, alias, context, resultMode, simple) {
        this.addMethod(method, obj, alias, context, resultMode, true, simple);
    }

    this.addMethods = function(methods, obj, aliases, context, resultMode, async, simple) {
        var count = methods.length;
        var i;
        if (aliases === undefined || aliases == null) {
            for (i = 0; i < count; i++) {
                this.addMethod(methods[i], obj, null, context, resultMode, async, simple);
            }
        }
        else {
            if (count != aliases.length) {
                throw new HproseException('The count of methods is not matched with aliases');
            }
            for (i = 0; i < count; i++) {
                this.addMethod(methods[i], obj, aliases[i], context, resultMode, async, simple);
            }
        }
    }

    this.addAsyncMethods = function(methods, obj, aliases, context, resultMode, simple) {
        this.addMethods(methods, obj, aliases, context, resultMode, true, simple);
    }

    this.addInstanceMethods = function(obj, aliasPrefix, context, resultMode, async, simple) {
        var alias;
        for (var name in obj) {
            alias = (aliasPrefix ? aliasPrefix + "_" + name : name);
            if (typeof(obj[name]) == 'function') {
                this.addMethod(obj[name], obj, alias, context, resultMode, async, simple);
            }
        }
    }

    this.addAsyncInstanceMethods = function(obj, aliasPrefix, context, resultMode, async, simple) {
        this.addInstanceMethods(obj, aliasPrefix, context, resultMode, true, simple);
    }

    this.getFilter = function() {
        return m_filter;
    }

    this.setFilter = function(filter) {
        m_filter = filter;
    }
}

util.inherits(HproseService, EventEmitter);

module.exports = HproseService;

File: src/dotnet/src/Hprose/Server/HproseService.cs
/**********************************************************\
|                                                          |
|                          hprose                          |
|                                                          |
| Official WebSite: http://www.hprose.com/                 |
|                   http://www.hprose.net/                 |
|                   http://www.hprose.org/                 |
|                                                          |
\**********************************************************/
/**********************************************************\
 *                                                        *
 * HproseService.cs                                       *
 *                                                        *
 * hprose service class for C#.                           *
 *                                                        *
 * LastModified: Feb 18, 2014                             *
 * Author: Ma Bingyao <andot@hprose.com>                  *
 *                                                        *
\**********************************************************/
#if !ClientOnly
using System;
using System.Collections;
#if !(dotNET10 || dotNET11 || dotNETCF10)
using System.Collections.Generic;
#endif
using System.IO;
using System.Reflection;
using Hprose.IO;
using Hprose.Common;

namespace Hprose.Server {

    public abstract class HproseService {

        private HproseMode mode = HproseMode.FieldMode;
        private bool debugEnabled = false;
        protected HproseMethods gMethods = null;
        public event BeforeInvokeEvent OnBeforeInvoke = null;
        public event AfterInvokeEvent OnAfterInvoke = null;
        public event SendErrorEvent OnSendError = null;
        private IHproseFilter filter = null;

        public virtual HproseMethods GlobalMethods {
            get {
                if (gMethods == null) {
                    gMethods = new HproseMethods();
                }
                return gMethods;
            }
            set {
                gMethods = value;
            }
        }

        public HproseMode Mode {
            get {
                return mode;
            }
            set {
                mode = value;
            }
        }

        public bool IsDebugEnabled {
            get {
                return debugEnabled;
            }
            set {
                debugEnabled = value;
            }
        }

        public IHproseFilter Filter {
            get {
                return filter;
            }
            set {
                filter = value;
            }
        }

        public void Add(MethodInfo method, object obj, string aliasName) {
            GlobalMethods.AddMethod(method, obj, aliasName);
        }

        public void Add(MethodInfo method, object obj, string aliasName, HproseResultMode mode) {
            GlobalMethods.AddMethod(method, obj, aliasName, mode);
        }

        public void Add(string methodName, object obj, Type[] paramTypes, string aliasName) {
            GlobalMethods.AddMethod(methodName, obj, paramTypes, aliasName);
        }

        public void Add(string methodName, object obj, Type[] paramTypes, string aliasName, HproseResultMode mode) {
            GlobalMethods.AddMethod(methodName, obj, paramTypes, aliasName, mode);
        }

        public void Add(string methodName, Type type, Type[] paramTypes, string aliasName) {
            GlobalMethods.AddMethod(methodName, type, paramTypes, aliasName);
        }

        public void Add(string methodName, Type type, Type[] paramTypes, string aliasName, HproseResultMode mode) {
            GlobalMethods.AddMethod(methodName, type, paramTypes, aliasName, mode);
        }

        public void Add(string methodName, object obj, Type[] paramTypes) {
            GlobalMethods.AddMethod(methodName, obj, paramTypes);
        }

        public void Add(string methodName, object obj, Type[] paramTypes, HproseResultMode mode) {
            GlobalMethods.AddMethod(methodName, obj, paramTypes, mode);
        }

        public void Add(string methodName, Type type, Type[] paramTypes) {
            GlobalMethods.AddMethod(methodName, type, paramTypes);
        }

        public void Add(string methodName, Type type, Type[] paramTypes, HproseResultMode mode) {
            GlobalMethods.AddMethod(methodName, type, paramTypes, mode);
        }

        public void Add(string methodName, object obj, string aliasName) {
            GlobalMethods.AddMethod(methodName, obj, aliasName);
        }

        public void Add(string methodName, object obj, string aliasName, HproseResultMode mode) {
            GlobalMethods.AddMethod(methodName, obj, aliasName, mode);
        }

        public void Add(string methodName, Type type, string aliasName) {
            GlobalMethods.AddMethod(methodName, type, aliasName);
        }

        public void Add(string methodName, Type type, string aliasName, HproseResultMode mode) {
            GlobalMethods.AddMethod(methodName, type, aliasName, mode);
        }

        public void Add(string methodName, object obj) {
            GlobalMethods.AddMethod(methodName, obj);
        }

        public void Add(string methodName, object obj, HproseResultMode mode) {
            GlobalMethods.AddMethod(methodName, obj, mode);
        }

        public void Add(string methodName, Type type) {
            GlobalMethods.AddMethod(methodName, type);
        }

        public void Add(string methodName, Type type, HproseResultMode mode) {
            GlobalMethods.AddMethod(methodName, type, mode);
        }

        public void Add(string[] methodNames, object obj, string[] aliasNames) {
            GlobalMethods.AddMethods(methodNames, obj, aliasNames);
        }

        public void Add(string[] methodNames, object obj, string[] aliasNames, HproseResultMode mode) {
            GlobalMethods.AddMethods(methodNames, obj, aliasNames, mode);
        }

        public void Add(string[] methodNames, object obj, string aliasPrefix) {
            GlobalMethods.AddMethods(methodNames, obj, aliasPrefix);
        }

        public void Add(string[] methodNames, object obj, string aliasPrefix, HproseResultMode mode) {
            GlobalMethods.AddMethods(methodNames, obj, aliasPrefix, mode);
        }

        public void Add(string[] methodNames, object obj) {
            GlobalMethods.AddMethods(methodNames, obj);
        }

        public void Add(string[] methodNames, object obj, HproseResultMode mode) {
            GlobalMethods.AddMethods(methodNames, obj, mode);
        }

        public void Add(string[] methodNames, Type type, string[] aliasNames) {
            GlobalMethods.AddMethods(methodNames, type, aliasNames);
        }

        public void Add(string[] methodNames, Type type, string[] aliasNames, HproseResultMode mode) {
            GlobalMethods.AddMethods(methodNames, type, aliasNames, mode);
        }

        public void Add(string[] methodNames, Type type, string aliasPrefix) {
            GlobalMethods.AddMethods(methodNames, type, aliasPrefix);
        }

        public void Add(string[] methodNames, Type type, string aliasPrefix, HproseResultMode mode) {
            GlobalMethods.AddMethods(methodNames, type, aliasPrefix, mode);
        }

        public void Add(string[] methodNames, Type type) {
            GlobalMethods.AddMethods(methodNames, type);
        }

        public void Add(string[] methodNames, Type type, HproseResultMode mode) {
            GlobalMethods.AddMethods(methodNames, type, mode);
        }

        public void Add(object obj, Type type, string aliasPrefix) {
            GlobalMethods.AddInstanceMethods(obj, type, aliasPrefix);
        }

        public void Add(object obj, Type type, string aliasPrefix, HproseResultMode mode) {
            GlobalMethods.AddInstanceMethods(obj, type, aliasPrefix, mode);
        }

        public void Add(object obj, Type type) {
            GlobalMethods.AddInstanceMethods(obj, type);
        }

        public void Add(object obj, Type type, HproseResultMode mode) {
            GlobalMethods.AddInstanceMethods(obj, type, mode);
        }

        public void Add(object obj, string aliasPrefix) {
            GlobalMethods.AddInstanceMethods(obj, aliasPrefix);
        }

        public void Add(object obj, string aliasPrefix, HproseResultMode mode) {
            GlobalMethods.AddInstanceMethods(obj, aliasPrefix, mode);
        }

        public void Add(object obj) {
            GlobalMethods.AddInstanceMethods(obj);
        }

        public void Add(object obj, HproseResultMode mode) {
            GlobalMethods.AddInstanceMethods(obj, mode);
        }

        public void Add(Type type, string aliasPrefix) {
            GlobalMethods.AddStaticMethods(type, aliasPrefix);
        }

        public void Add(Type type, string aliasPrefix, HproseResultMode mode) {
            GlobalMethods.AddStaticMethods(type, aliasPrefix, mode);
        }

        public void Add(Type type) {
            GlobalMethods.AddStaticMethods(type);
        }

        public void Add(Type type, HproseResultMode mode) {
            GlobalMethods.AddStaticMethods(type, mode);
        }

        public void AddMissingMethod(string methodName, object obj) {
            GlobalMethods.AddMissingMethod(methodName, obj);
        }

        public void AddMissingMethod(string methodName, object obj, HproseResultMode mode) {
            GlobalMethods.AddMissingMethod(methodName, obj, mode);
        }

        public void AddMissingMethod(string methodName, Type type) {
            GlobalMethods.AddMissingMethod(methodName, type);
        }

        public void AddMissingMethod(string methodName, Type type, HproseResultMode mode) {
            GlobalMethods.AddMissingMethod(methodName, type, mode);
        }

        private void ResponseEnd(Stream ostream, MemoryStream data, string error) {
            if (filter != null) ostream = filter.OutputFilter(ostream);
            if (error != null && OnSendError != null) {
                OnSendError(error);
            }
            data.WriteTo(ostream);
            ostream.Flush();
        }

        protected virtual object[] FixArguments(Type[] argumentTypes, object[] arguments, int count) {
            return arguments;
        }

        protected void SendError(Stream ostream, string error) {
            MemoryStream data = new MemoryStream(4096);
            HproseWriter writer = new HproseWriter(data, mode);
            data.WriteByte(HproseTags.TagError);
            writer.WriteString(error);
            data.WriteByte(HproseTags.TagEnd);
            ResponseEnd(ostream, data, error);
        }

        protected void DoInvoke(Stream istream, Stream ostream, HproseMethods methods) {
            HproseReader reader = new HproseReader(istream, mode);
            MemoryStream data = new MemoryStream(4096);
            HproseWriter writer = new HproseWriter(data, mode);
            int tag;
            do {
                reader.Reset();
                string name = reader.ReadString();
                HproseMethod remoteMethod = null;
                int count = 0;
                object[] args = null;
                object[] arguments = null;
                bool byRef = false;
                tag = reader.CheckTags((char)HproseTags.TagList + "" +
                                       (char)HproseTags.TagEnd + "" +
                                       (char)HproseTags.TagCall);
                if (tag == HproseTags.TagList) {
                    reader.Reset();
                    count = reader.ReadInt(HproseTags.TagOpenbrace);
                    if (methods != null) {
                        remoteMethod = methods.GetMethod(name, count);
                    }
                    if (remoteMethod == null) {
                        remoteMethod = GlobalMethods.GetMethod(name, count);
                    }
                    if (remoteMethod == null) {
                        arguments = reader.ReadArray(count);
                    }
                    else {
                        arguments = new object[count];
                        reader.ReadArray(remoteMethod.paramTypes, arguments, count);
                    }
                    tag = reader.CheckTags((char)HproseTags.TagTrue + "" +
                                           (char)HproseTags.TagEnd + "" +
                                           (char)HproseTags.TagCall);
                    if (tag == HproseTags.TagTrue) {
                        byRef = true;
                        tag = reader.CheckTags((char)HproseTags.TagEnd + "" +
                                               (char)HproseTags.TagCall);
                    }
                }
                else {
                    if (methods != null) {
                        remoteMethod = methods.GetMethod(name, 0);
                    }
                    if (remoteMethod == null) {
                        remoteMethod = GlobalMethods.GetMethod(name, 0);
                    }
                    arguments = new object[0];
                }
                if (OnBeforeInvoke != null) {
                    OnBeforeInvoke(name, arguments, byRef);
                }
                if (remoteMethod == null) {
                    args = arguments;
                }
                else {
                    args = FixArguments(remoteMethod.paramTypes, arguments, count);
                }
                object result;
                if (remoteMethod == null) {
                    if (methods != null) {
                        remoteMethod = methods.GetMethod("*", 2);
                    }
                    if (remoteMethod == null) {
                        remoteMethod = GlobalMethods.GetMethod("*", 2);
                    }
                    if (remoteMethod == null) {
                        throw new MissingMethodException("Can't find this method " + name);
                    }
                    result = remoteMethod.method.Invoke(remoteMethod.obj, new object[] { name, args });
                }
                else {
                    result = remoteMethod.method.Invoke(remoteMethod.obj, args);
                }
                if (byRef) {
                    Array.Copy(args, 0, arguments, 0, count);
                }
                if (OnAfterInvoke != null) {
                    OnAfterInvoke(name, arguments, byRef, result);
                }
                if (remoteMethod.mode == HproseResultMode.RawWithEndTag) {
                    data.Write((byte[])result, 0, ((byte[])result).Length);
                    ResponseEnd(ostream, data, null);
                    return;
                }
                else if (remoteMethod.mode == HproseResultMode.Raw) {
                    data.Write((byte[])result, 0, ((byte[])result).Length);
                }
                else {
                    data.WriteByte(HproseTags.TagResult);
                    if (remoteMethod.mode == HproseResultMode.Serialized) {
                        data.Write((byte[])result, 0, ((byte[])result).Length);
                    }
                    else {
                        writer.Reset();
                        writer.Serialize(result);
                    }
                    if (byRef) {
                        data.WriteByte(HproseTags.TagArgument);
                        writer.Reset();
                        writer.WriteArray(arguments);
                    }
                }
            } while (tag == HproseTags.TagCall);
            data.WriteByte(HproseTags.TagEnd);
            ResponseEnd(ostream, data, null);
        }

        protected void DoFunctionList(Stream ostream, HproseMethods methods) {
#if !(dotNET10 || dotNET11 || dotNETCF10)
            List<string> names = new List<string>(GlobalMethods.AllNames);
#else
            ArrayList names = new ArrayList(GlobalMethods.AllNames);
#endif
            if (methods != null) {
                names.AddRange(methods.AllNames);
            }
            MemoryStream data = new MemoryStream(4096);
            HproseWriter writer = new HproseWriter(data, mode);
            data.WriteByte(HproseTags.TagFunctions);
#if !(dotNET10 || dotNET11 || dotNETCF10)
            writer.WriteList((IList<string>)names);
#else
            writer.WriteList((IList)names);
#endif
            data.WriteByte(HproseTags.TagEnd);
            ResponseEnd(ostream, data, null);
        }

        public void Handle(Stream istream, Stream ostream, HproseMethods methods) {
            try {
                if (filter != null) istream = filter.InputFilter(istream);
                int tag = istream.ReadByte();
                switch (tag) {
                    case HproseTags.TagCall:
                        DoInvoke(istream, ostream, methods);
                        break;
                    case HproseTags.TagEnd:
                        DoFunctionList(ostream, methods);
                        break;
                    default:
                        SendError(ostream, "Unknown Tag");
                        break;
                }
            }
            catch (Exception e) {
                if (debugEnabled) {
                    SendError(ostream, e.ToString());
                }
                else {
                    SendError(ostream, e.Message);
                }
            }
        }
    }
}
#endif